### ISSUE TITLE
Visualizing the Attention Masks

### ISSUE THREAD

**Issue by Albert337** [haoxintong/ResidualAttentionNetwork-Gluon#1]
Hi,I wanted to ask how to visualize the attention masks as given in Fig 1 in the paper? Does it involve using GRAD CAM? or is it directly the actual outputs of the mask? Also, given that the masks are of different sizes and the number of channels are more (usually 256, 512 or 1024), how do the authors visualize the attention masks then?and if i use it on my owner data,how to fix the code?
A code snippet would help highly.

**Comment by haoxintong** [comment]
Hi, 
In my experiment, I use the code below to get the mask branches output :
```python

...
...
outputs = [symbols["attentionnet0_attentionblock%d_mask_sigmoid0_fwd_output" % i] for i in range(3)]
...

mask1, mask2, mask3 = net(img)
for i in range(len(m1[0])):
       mask = mask1[0][i:i + 1].asnumpy()
       mask = mask.transpose((1, 2, 0))

       img = np.multiply(img, mask)
       img = np.array(img, dtype=np.uint8)
```
But I have no idea of how to get the same result of Fig 1 in paper so far. For different images the output channels are not similar, maybe I should compute the max activation value of the channels and use this channel as the  mask.

**Comment by Albert337** [comment]
My computor environment is nvidia GTX1050, ubuntn16.04(only one gpu),when i run your code, i edit your code where num-gpu to 1,but an error occured that is gpu not enabled error.when i set it to zero,the code is working with cpu,but the speed is slowly.To work with gpu how to edit it ?thanks a lot!Have you tried the attention model coding with tensorflow?

**Comment by haoxintong** [comment]
Did you use mxnet gpu version? `pip install mxnet` will only provide cpu training, `pip install mxnet-cu80` is for mxnet training on gpu with cuda-8.0. And as the gpu memory of GTX1050 is lower, the batch size should be set 32 or 16 for attention-cifar10. 
Tensorflow is not convenient for coding，pytorch or mxnet-gluon is better choice for experiments.

**Comment by Albert337** [comment]
Thanks! I get it.
Em, I really appreciate your works ,espically the results of your experiments,
but I am not familiar with mxnet, would you mind adding a friend?my email is xiaoxiaoamugong@163.com .

**Comment by haoxintong** [comment]
For mxnet resources, it is strongly recommended to go [Gluon forum](https://discuss.gluon.ai/), and [Dive Into DL](https://zh.diveintodeeplearning.org/) tutorials. In this chinese community you will learn anything you want to know about mxnet and gluon in deep learning.

**Comment by Albert337** [comment]
Thanks a lot! Are you a Chinese?Have you trained and tested the model on your own data? And if i want to use the model to train on my own data,how to do it?



At 2018-12-18 09:40:05, "Maybe" <notifications@github.com> wrote:


For mxnet resources, it is strongly recommended to go Gluon forum, and Dive Into DL tutorials. In this chinese community you will learn anything you want to know about mxnet and gluon in deep learning.

**Comment by haoxintong** [comment]
后边沟通我就用中文说吧.
res-attention训练都比较吃显存, 如果资源充足的话, 就按照正常的训练流程走就可以了.
比如说数据读取-预处理-切分batch到多卡-训练更新参数-测试. 

我自己跑的话, 就只做过几个toy-dataset的测试, 毕竟训练资源紧缺.

**Comment by Albert337** [comment]
当我配置好mxnet-cu90(因为我以前用的是cuda9.0)后,运行程序出现了
![default](https://user-images.githubusercontent.com/33714840/50154257-cc5ddf00-0303-11e9-8d22-41431e7c6b9d.png),我不太清楚为什么你在数据扩容函数image.CreateAugmenter这里mean=nd.array([0.485, 0.456, 0.406]),std=nd.array([0.229, 0.224, 0.225])这些定值是怎么来的额?出现我那种错误是什么原因呢?

**Comment by haoxintong** [comment]
这些是通过imagenet数据计算出来的，一般认为在别的地方也适用，毕竟做迁移学习的时候通常用的也是在imagenet上预训练的模型。
至于你这个错误，我的mxnet版本这句话是这么写的：
```python
assert isinstance(mean, (np.ndarray, nd.NDArray)) and mean.shape[0] in [1, 3]
```
所以两种解决方法，一个是升级mxnet版本，另一个是把输入中的`nd.array()`换成`np.array()`

**Comment by Albert337** [comment]
这个问题我已经解决了,想问一下你自己测试自己数据的代码还在么?我想导入自己的数据需要先转化成什么格式的?还是直接导入也可以?

**Comment by haoxintong** [comment]
这个用标准的读入方法就行, 具体参照上边给你的中文教程中的例子

**Comment by Albert337** [comment]
你好,不好意思再次打扰你了!我已经将我的数据转化为rec格式,可是我不知道怎么在你的程序上进行微调导入我自己的数据了?你有什么好的推荐么?或许你以前是怎么测试自己数据集的额,函数接口还在么?

**Comment by haoxintong** [comment]
参考这个[ImageRecordDataset](http://mxnet.incubator.apache.org/api/python/gluon/data.html#mxnet.gluon.data.vision.datasets.ImageRecordDataset)用法，也可以看[这个类的实现](https://github.com/THUFutureLab/gluon-fr/blob/master/gluonfr/data/dataset.py#L36)，都是处理rec的，只不过根据制作rec流程的不同，读取时候要做些调整

**Comment by Albert337** [comment]
![default](https://user-images.githubusercontent.com/33714840/50323533-0eaa3a80-0515-11e9-8f9e-5c50fff464d1.png)
我利用ImageRecordDataset读取自己的数据,但是再导入网络net(x)这里开始报错初始化问题,这种问题是我导入的数据不对么?(我尝试print过,与cifar_10的格式差不多)

**Comment by haoxintong** [comment]
这个问题是你的网络没做初始化, 训练开始前要调用`net.initialize()`

**Comment by Albert337** [comment]
![default](https://user-images.githubusercontent.com/33714840/50328245-589f1a80-052d-11e9-8ce0-7db708bf8a81.png)你的意思是在这里将net.initialize(init=mx.initializer.MSRAPrelu(), ctx=ctx)修改一下?

**Comment by haoxintong** [comment]
我漏看了你的数据输入, 这个初始化是这么写没错. 实际问题是我写的cifar net是为了(batch_size, 3, 32, 32)大小输入设计网络的, 所以当更改输入尺寸到(160, 160)时, 需要在网络里修改mask分支的输入输出尺寸. 所以解决这个问题也是两种方法, 一个是修改输入大小变成(224,224)之后调用get_attention_net(), 另一种是改一版适用你的图像大小的网络结构:
```python
       self.features = nn.HybridSequential()
            # 32x32
            self.features.add(nn.Conv2D(32, 3, 1, 1, use_bias=False))
            self.features.add(nn.BatchNorm())
            self.features.add(nn.Activation('relu'))

            # 32x32
            self.features.add(BottleneckV2(128, 1, True, 32))
            for _ in range(modules[0]):
                self.features.add(AttentionBlock(128, 32, 1, p, t, r)) # 这里第二个参数要和输入的特征图尺寸相同

            # 16x16
            self.features.add(BottleneckV2(256, 2, True, 128))
            for _ in range(modules[1]):
                self.features.add(AttentionBlock(256, 16, 2, p, t, r))

            # 8x8
            self.features.add(BottleneckV2(512, 2, True, 256))
            for _ in range(modules[2]):
                self.features.add(AttentionBlock(512, 8, 3, p, t, r))

            # 8x8
            self.features.add(BottleneckV2(1024, 2, True, 512),
                              BottleneckV2(1024, 1),
                              BottleneckV2(1024, 1))

            # 2048
            self.features.add(nn.BatchNorm(),
                              nn.Activation('relu'),
                              nn.GlobalAvgPool2D(),
                              nn.Flatten())

```

**Comment by Albert337** [comment]
大神,你这网络结构我读的不是很明白?你的trunck branch和mask branch在哪里结合的额?我现在想修改网络结构使用我自己的图片,在你注释那里我修改成160,还是会报同样的错,我这还需要在哪里修改下额?

**Comment by haoxintong** [comment]
理论上来讲改成这样就能跑了，但是这个结构未必是精度最好。上边只给你注释了一个， 要改所有Attention模块都得改。trunk 和 mask的结合看attentionblock的实现。
```python
   self.features = nn.HybridSequential()
            # 160x160
            self.features.add(nn.Conv2D(32, 3, 2, 1, use_bias=False))
            self.features.add(nn.BatchNorm())
            self.features.add(nn.Activation('relu'))

            # 80x80
            self.features.add(nn.MaxPool2D(pool_size=3, strides=2, padding=1))

           # 40x40
            self.features.add(BottleneckV2(128, 1, True, 32))
            for _ in range(modules[0]):
                self.features.add(AttentionBlock(128, 40, 1, p, t, r)) 

            # 20x20
            self.features.add(BottleneckV2(256, 2, True, 128))
            for _ in range(modules[1]):
                self.features.add(AttentionBlock(256, 20, 2, p, t, r))

            # 10x10
            self.features.add(BottleneckV2(512, 2, True, 256))
            for _ in range(modules[2]):
                self.features.add(AttentionBlock(512, 10, 3, p, t, r))

            # 10x10
            self.features.add(BottleneckV2(1024, 2, True, 512),
                              BottleneckV2(1024, 1),
                              BottleneckV2(1024, 1))

            # 2048
            self.features.add(nn.BatchNorm(),
                              nn.Activation('relu'),
                              nn.GlobalAvgPool2D(),
                              nn.Flatten())
```

**Comment by Albert337** [comment]
我按照你说的进行修改了,还是报原来的错误,看来网络结构还需要修改?

**Comment by Albert337** [comment]
我查看了输入的数据data
![default](https://user-images.githubusercontent.com/33714840/50380420-6f667e00-06a1-11e9-86bb-97d7a4d2c550.png),网络输入的X就是四维data的图片数据,这个输入应该也没问题,....

**Comment by Albert337** [comment]
大神你好，我把你的代码再次仔细看了下，你的整个项目下有六个文件夹，data里放了5张结果图，log和models文件夹用于保存训练模型参数，net文件夹下有3个py文件（一个空的初始化py文件，一个attention_block.py,一个attention_net.py），scripts文件存下sh文件放了输入的命令脚本，utils下有两个py文件（一个cifar_dataset.py,一个reorg_cifar10_data.py）,接下来剩下cifar10_train.py和cifar10_test.py文件。
通过对cifar10_train.py文件内容的理解，你通过gluon.data.vision.CIFAR10导入cifar10数据，DataLoader将cifar10数据分批，然后在训练，好像没用到utils里对cifar10数据的处理。神经网络通过调用attention_net里的get_attention_cifar（）函数（貌似get_attention_net等函数没有用上），而get_attention_cifar（）函数又调用了AttentionNetCifar类，在该类的构造函数内又调用了mxnet.gluon.model_zoo.vision.resnet中的BottleneckV2来构建残差单元，调用了attention_block中的AttentionBlock类，通过类内的函数hybrid_forward（这个函数会自动调用么？）将mask branch和trunk branch结合在一起来搭建residual attention network。
感觉mxnet.gluon编码和Keras风格很像，还是没找到修改方案，还希望大神可以抽空帮我看看呗（大神上面说的mask branches output我好像在代码中也没看到，原谅我这小菜鸡太弱太弱。。。）~~~

**Comment by haoxintong** [comment]
uitls那里边的东西没什么用，就是写着练手的，可以参考那些代码来写自己的数据IO类。

上边的修改我在本机测试是没问题的，
```python
outputs = net(mx.nd.ones(shape=(32, 3, 160, 160), ctx=mx.gpu()))
```
这行直接能过，你这个可以检查一下data[0].shape，如果这个正确再看看是不是网络没有调用到你修改过的部分。

gluon的风格其实和pytorch的API设计在代码上几乎一样，框架设计上的区别只有深入了解的时候才能感受到。像上边说的获取mask输出，其实用的还是mxnet symbol的那一套接口，要先序列化再导入到gluon的SymbolBlock。 如果纯用gluon的方式就需要在网络结构中显式的输出，然后一层一层在外边接这个结果。

刚接触gluon这个API， 可以先从最小的结构和例子理解，后面大的网络就是这些结构的堆叠了，参考[Block用法](https://zh.diveintodeeplearning.org/chapter_deep-learning-computation/model-construction.html)和[HybridBlock用法](https://zh.diveintodeeplearning.org/chapter_computational-performance/hybridize.html)。

**Comment by Albert337** [comment]
我看了下data[0].shape=[32,3,160,160],调用的网络修改主要在attentionnetcifar类的构造函数内,还是出现未初始化问题,我在简书上看到一个人出现和我一样的问题,https://www.jianshu.com/p/ecd46240c6ba,他说是由于mxnet/gluon的特性导致的,可他的解决方案不是很会操作.....

**Comment by haoxintong** [comment]
我看了一下简书的文章，对你这个情况应该没什么用。延后初始化机制是因为在写网络结构时免去具体的输入输出形状推理，只根据输入就能确定参数形状，但仍然要按顺序调用`net=Net()`、`net.initialize()`、`net(X)`这些标准流程，之后才能访问参数。具体的介绍参考[这个](https://zh.diveintodeeplearning.org/chapter_deep-learning-computation/deferred-init.html)。

我跑这段[代码](https://gist.github.com/haoxintong/3bce274b52e289b0c8c00c96fff30477)是正常的， 你看看和你修改的实现有没有什么区别。

**Comment by Albert337** [comment]
不好意思，你代码的链接打不开,可以把它贴出来么？

**Comment by haoxintong** [comment]
```python
import mxnet as mx
from mxnet.gluon import nn
from net.attention_block import BottleneckV2, AttentionBlock


class AttentionNet160(nn.HybridBlock):
    def __init__(self, classes, modules, p, t, r, **kwargs):
        super().__init__(**kwargs)
        assert len(modules) == 3
        with self.name_scope():
            self.features = nn.HybridSequential()
            # 160x160
            self.features.add(nn.Conv2D(32, 3, 2, 1, use_bias=False))
            self.features.add(nn.BatchNorm())
            self.features.add(nn.Activation('relu'))

            # 80x80
            self.features.add(nn.MaxPool2D(pool_size=3, strides=2, padding=1))

            # 40x40
            self.features.add(BottleneckV2(128, 1, True, 32))
            for _ in range(modules[0]):
                self.features.add(AttentionBlock(128, 40, 1, p, t, r))

                # 20x20
            self.features.add(BottleneckV2(256, 2, True, 128))
            for _ in range(modules[1]):
                self.features.add(AttentionBlock(256, 20, 2, p, t, r))

            # 10x10
            self.features.add(BottleneckV2(512, 2, True, 256))
            for _ in range(modules[2]):
                self.features.add(AttentionBlock(512, 10, 3, p, t, r))

            # 10x10
            self.features.add(BottleneckV2(1024, 2, True, 512),
                              BottleneckV2(1024, 1),
                              BottleneckV2(1024, 1))

            # 2048
            self.features.add(nn.BatchNorm(),
                              nn.Activation('relu'),
                              nn.GlobalAvgPool2D(),
                              nn.Flatten())
            self.output = nn.Dense(classes)

    def hybrid_forward(self, F, x, *args, **kwargs):
        x = self.features(x)
        x = self.output(x)
        return x


if __name__ == '__main__':

    net = AttentionNet160(4, (1, 1, 1), p=1, t=2, r=1)
    net.initialize(ctx=mx.gpu())
    net.hybridize()
    outputs = net(mx.nd.ones(shape=(32, 3, 160, 160), ctx=mx.gpu()))
```

**Comment by Albert337** [comment]
非常谢谢,程序已经调通了,可是我想在里面加上          
`            # Log the first batch of images of each epoch
            if i == 0:
               sw.add_image('data_batch', data[0].reshape((args.batch_size, 3, 160, 160)), epoch)和'sw.add_histogram(tag=name, values=grads[i], global_step=epoch, bins=1000)'等结果写入文件出错了....`

**Comment by haoxintong** [comment]
mxboard除了最基本的loss记录别的我也没用过, 你看代码里的注释说明排坑吧

**Comment by Albert337** [comment]
哦,还是非常感谢你,小姐姐!

**Comment by Albert337** [comment]
小姐姐,你真的不想亲眼目睹一下这个可视化过程,解决一下
![default](https://user-images.githubusercontent.com/33714840/50891352-091a8300-1437-11e9-9e65-3edd74794095.png)问题么?我在mxnet论坛上问了半天也没人能帮我解决....
